servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java

```diff
@@ -213,47 +202,38 @@ private static String absoluteToRelativeFormRequestTarget(final String requestTa
             final int questionMarkIdx = requestTarget.indexOf('?', fromIndex);
             return questionMarkIdx < 0 ? "/" : '/' + requestTarget.substring(questionMarkIdx);
         }
-
-        @Override
-        public Completable closeAsync() {
-            // Make a best effort to clear the map. Note that we don't attempt to resolve race conditions between
-            // closing the client and in flight requests adding Keys to the map. We also don't attempt to remove
-            // from the map if a request fails, or a request is made after the client is closed.
-            return new SubscribableCompletable() {
-                @Override
-                protected void handleSubscribe(final Subscriber subscriber) {
-                    urlKeyCache.clear();
-                    deliverCompleteFromSource(subscriber);
-                }
-            };
-        }
     }
 
     private static final class UrlKey {
         final String scheme;
-        final HostAndPort hostAndPort;
+        final String host;
+        final int port;
+        private final int hashCode;
 
-        UrlKey(final String scheme, final HostAndPort hostAndPort) {
+        UrlKey(final String scheme, final String host, final int port) {
             this.scheme = scheme;
-            this.hostAndPort = hostAndPort;
+            this.host = host;
+            this.port = port;
+            // hashCode is required at least once, but may be necessary multiple times for a single selectClient run
+            this.hashCode = 31 * (caseInsensitiveHashCode(host) + port) + scheme.hashCode();
         }
 
+        /**
+         * This class is used only within controlled scope, a non-null argument and class match are guaranteed.
+         */
+        @SuppressWarnings("EqualsWhichDoesntCheckParameterClass")
         @Override
         public boolean equals(final Object o) {
             if (this == o) {
                 return true;
             }
-            if (o == null || getClass() != o.getClass()) {
-                return false;
-            }
-
             final UrlKey urlKey = (UrlKey) o;
-            return scheme.equals(urlKey.scheme) && hostAndPort.equals(urlKey.hostAndPort);
+            return port == urlKey.port && scheme.equals(urlKey.scheme) && host.equalsIgnoreCase(urlKey.host);
```

Comment:
ffti: in my minds eye, the optimal ordering of checks would be `port -> host -> scheme`. It would seem uncommon for the host and port to match but not the scheme other than in TLS-detection scenarios which are uncommon and typically transient. I imagine in real scenarios the host will be the overwhelmingly highest cardinality of the three, but it's so cheap to check the port that it makes sense to do it first.

Reply:
Because the scheme comes from `Uri3986`, it's always a trivial ref check for http and https. I changed the order anyway, bcz agreed that scheme has less variance.